src/layout.cpp

```diff
@@ -0,0 +1,408 @@
+    if (exist) {
+      return true;
+    }
+  }
+  usedId[node.id] = false;
+  return false;
+}
+
+bool cycleExistsDFS(std::vector<TreeNode> &nodes) {
+  if (nodes.size() == 0) {
+      return false;
+  }
+    
+  std::vector<bool> usedId(nodes.size(), false);
+
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].pred.size() == 0) {
+      bool exist = algorithmDFS(nodes[i], nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  return algorithmDFS(nodes[0], nodes, usedId);
+}
+
+struct EdgeCount {
+  int degOut = 0;
+  int degIn = 0;
+};
+
+void removeEdgeCount(int i, std::vector<EdgeCount> &edgeCounts,
+                     std::vector<TreeNode> &nodes) {
+  const TreeNode &node = nodes[i];
+  for (TreeNode::nodeId id: node.succ) {
+    edgeCounts[id].degIn--;
+  }
+  for (TreeNode::nodeId id: node.pred) {
+    edgeCounts[id].degOut--;
+  }
+
+  edgeCounts[i].degIn = -1;
+  edgeCounts[i].degOut = -1;
+}
+
+bool comparatorId(TreeNode &first, TreeNode &second) {
+    return first.id < second.id;
+}
+
+void greedyFAS(
+    std::vector<TreeNode> &nodes,
+    std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges) {
+  std::vector<TreeNode::nodeId> s1 = {}, s2 = {};
+  
+    std::vector<EdgeCount> edgeCounts;
+    edgeCounts.reserve(nodes.size());
+    for (const TreeNode &node: nodes) {
+        edgeCounts.push_back({ static_cast<int>(node.succ.size()), static_cast<int>(node.pred.size()) });
+    }
+    
+  size_t lenNodes = nodes.size();
+  while (lenNodes > 0) {
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degOut == 0) {
+        s2.push_back(i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
+
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degIn == 0) {
+        s1.push_back(i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
+
+    if (lenNodes > 0) {
+      int max;
+      for (int i = 0; i < edgeCounts.size(); i++) {
+        if (edgeCounts[i].degOut >= 0 && edgeCounts[i].degIn >= 0) {
+          max = i;
+          break;
+        }
+      }
+      for (int i = 0; i < edgeCounts.size(); i++) {
+        if (edgeCounts[i].degOut - edgeCounts[i].degIn > max &&
+            edgeCounts[i].degOut >= 0 && edgeCounts[i].degIn >= 0) {
+          max = i;
+        }
+      }
+
+      s1.push_back(max);
+      removeEdgeCount(max, edgeCounts, nodes);
+
+      lenNodes--;
+    }
+  }
+
+    std::vector<TreeNode::nodeId> &ordered = s1;
+    ordered.insert(ordered.end(), s2.rbegin(), s2.rend());
+
+    for (int i = 0; i < ordered.size(); i++) {
+        for (int j = 0; j < nodes.size(); j++) {
+            if (nodes[j].id == ordered[i]) {
+                std::swap(nodes[i], nodes[j]);
+            }
+        }
+    }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = nodes[i].succ.size() - 1; j >= 0; j--) {
```

Comment:
Do not touch `nodes`, at all, you already have the `ordered`.
```suggestion
  std::vector<TreeNode::nodeId> &ordered = s1;
  ordered.insert(ordered.end(), s2.rbegin(), s2.rend());

  for (auto id: ordered) {
    const TreeNode &node = nodes[id];
    for (int j = node.succ.size() - 1; j >= 0; j--) {
```